velox/exec/Spill.cpp

```diff
@@ -0,0 +1,171 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/Spill.h"
+#include "velox/common/file/FileSystems.h"
+
+namespace facebook::velox::exec {
+
+std::atomic<int32_t> SpillStream::ordinalCounter_;
+uint64_t SpillState::sequence_ = 0;
+
+void SpillInput::next(bool throwIfPastEnd) {
+  int32_t readBytes = std::min(input_->size() - offset_, buffer_->capacity());
+  setRange({buffer_->asMutable<uint8_t>(), readBytes, 0});
+  input_->pread(offset_, readBytes, buffer_->asMutable<char>());
+  offset_ += readBytes;
+}
+
+void SpillInput::seekp(Position position) {
+  auto target = std::get<1>(position);
+  auto bufferOffset = offset_ - current_->size;
+  if (bufferOffset <= target && bufferOffset + current_->size < target) {
+    current_->position = target - bufferOffset;
+  } else {
+    // The seek target is not in the buffer.
+    offset_ = target;
+    current_->position = 0;
+    current_->size = 0;
+    next(true);
+  }
+}
+
+SpillFile::~SpillFile() {
+  if (path_[0] == '/') {
+    if (unlink(path_.c_str()) != 0) {
+      LOG(ERROR) << "Error deleting spill file " << path_
+                 << " errno: " << errno;
+    }
+  }
+}
+
+WriteFile& SpillFile::output() {
+  if (!output_) {
+    auto fs = filesystems::getFileSystem(path_, nullptr);
+    output_ = fs->openFileForWrite(path_);
+  }
+  return *output_;
+}
+
+void SpillFile::startRead() {
+  constexpr uint64_t kMaxReadBufferSize = 1 << 20; // 1MB
+  VELOX_CHECK(!output_);
+  auto fs = filesystems::getFileSystem(path_, nullptr);
+  auto file = fs->openFileForRead(path_);
+  auto buffer = AlignedBuffer::allocate<char>(
+      std::min<uint64_t>(fileSize_, kMaxReadBufferSize), &pool_);
+  input_ = std::make_unique<SpillInput>(std::move(file), std::move(buffer));
+  next();
+  index_ = 0;
+}
+
+void SpillFileList::flush(bool close) {
+  std::string str;
+  {
+    if (batch_) {
+      std::stringstream out;
+      batch_->flush(&out);
+      batch_.reset();
+      str = out.str();
+      if (!isOpen_ || (!close && files_.back()->size() > targetFileSize_)) {
+        if (isOpen_) {
+          files_.back()->finishWrite();
+        }
+        files_.push_back(std::make_unique<SpillFile>(
+            type_, fmt::format("{}-{}", path_, files_.size()), pool_));
+        isOpen_ = true;
+      }
+      files_.back()->output().append(str);
+    }
+  }
+  if (close && isOpen_) {
+    files_.back()->finishWrite();
+    isOpen_ = false;
+  }
+}
+
+void SpillFileList::write(
+    RowVectorPtr rows,
+    const folly::Range<IndexRange*> indices) {
+  if (!batch_) {
+    batch_ = std::make_unique<VectorStreamGroup>(&mappedMemory_);
+    batch_->createStreamTree(
+        std::static_pointer_cast<const RowType>(rows->type()), 1000);
+  }
+  batch_->append(rows, indices);
+  flush(false);
```

Comment:
Why flush(false) ? Do you expect external callers only to call flush(true) ? In this case, it would be cleaner to add publish flush() or finishWrite() method without any arguments and make existing flush(bool) method private.